File: Lists/SortLists/SortLists.scala
object Solution {
    def sortList(head: ListNode): ListNode = {
        if ((head == null) || (head.next == null)) {
            return head
        }
        var fast = head.next
        var slow = head

        while ((fast != null) && (fast.next != null)){
            fast = fast.next.next
            slow = slow.next
        }

        var mid = slow.next
        slow.next = null
        var l = sortList(head)
        var r = sortList(mid)
        return merge(l, r)        
    }

    def merge(l: ListNode, r: ListNode): ListNode = {
        if (l == null) return r
        if (r == null) return l
        if (l.x > r.x) {
            r.next = merge(l, r.next)
            return r
        } else {
            l.next = merge(l.next, r)
            return l
        }
    }
}

File: Lists/SortLists/SortLists.py
from typing import Optional, ListNode
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head
        
        fast = head.next
        slow = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        mid = slow.next
        slow.next = None
        l = self.sortList(head)
        r = self.sortList(mid)

        return self.merge(l, r)

    def merge(self, left, right):
        if not left or not right:
            return left or right
        if left.val > right.val:
            right.next = self.merge(left, right.next)
            return right
        else:
            left.next = self.merge(left.next, right)
            return left